File: perl/test/test_csv.pl
#!/usr/bin/env perl
#
# test_csv.pl - Test MPC designation conversion against CSV test data
#
# Usage: perl test_csv.pl <csv_file>
#

use strict;
use warnings;
use FindBin;
use lib "$FindBin::Bin/../src";
use MPC::Designation qw(convert_simple);
use Time::HiRes qw(time);

my $total_tests = 0;
my $passed_tests = 0;
my $failed_tests = 0;
my @errors;
my $max_errors = 100;

sub run_tests {
    my ($csv_file) = @_;

    open my $fh, '<', $csv_file or die "Cannot open file: $csv_file\n";

    my $start_time = time();

    # Skip header
    <$fh>;

    while (my $line = <$fh>) {
        chomp $line;
        next unless $line;

        my @parts = split /,/, $line;
        next unless @parts >= 2;

        my $unpacked = $parts[0];
        my $expected_packed = $parts[1];

        $total_tests++;

        eval {
            my $got_packed = convert_simple($unpacked);
            if ($got_packed ne $expected_packed) {
                $failed_tests++;
                if (@errors < $max_errors) {
                    push @errors, [$unpacked, $got_packed, $expected_packed];
                }
            } else {
                $passed_tests++;
            }
        };
        if ($@) {
            $failed_tests++;
            my $err = $@;
            $err =~ s/\n$//;
            if (@errors < $max_errors) {
                push @errors, [$unpacked, "ERROR: $err", $expected_packed];
            }
        }

        if ($total_tests % 100000 == 0) {
            print "Processed $total_tests entries...\n";
        }
    }

    close $fh;

    my $elapsed = time() - $start_time;
    my $elapsed_ms = int($elapsed * 1000);
    my $rate = $total_tests > 0 ? $total_tests / $elapsed : 0;

    print "\n";
    print "=== Test Results ===\n";
    print "Total:  $total_tests\n";
    print "Passed: $passed_tests\n";
    print "Failed: $failed_tests\n";
    printf "Time:   %dms (%.1f entries/sec)\n", $elapsed_ms, $rate;

    if ($failed_tests > 0 && @errors) {
        print "\n=== First " . scalar(@errors) . " failures ===\n";
        for my $err (@errors) {
            printf "%-25s %-15s %-15s\n", @$err;
        }
    }
}

# Main
if (@ARGV < 1) {
    print STDERR "Usage: test_csv.pl <csv_file>\n";
    exit 1;
}

run_tests($ARGV[0]);
exit($failed_tests > 0 ? 1 : 0);

File: php/test/test_csv.php
#!/usr/bin/env php
<?php
/**
 * Test mpc_designation.php against CSV file of known conversions.
 *
 * Usage: php test_csv.php <csv_file> [max_errors]
 */

declare(strict_types=1);

require_once __DIR__ . '/../src/MPCDesignation.php';

use MPC\MPCDesignation;
use MPC\MPCDesignationException;

function runTests(string $csvFile, int $maxErrors = 100): bool {
    $total = 0;
    $passed = 0;
    $failed = 0;
    $errors = [];

    $startTime = microtime(true);

    $fp = fopen($csvFile, 'r');
    if ($fp === false) {
        fwrite(STDERR, "Error: Cannot open file: $csvFile\n");
        return false;
    }

    // Skip header line
    fgets($fp);

    while (($line = fgets($fp)) !== false) {
        $line = trim($line);
        if ($line === '') {
            continue;
        }

        // Skip header if encountered
        if (str_starts_with($line, 'unpacked')) {
            continue;
        }

        $total++;
        $parts = explode(',', $line);
        $unpacked = $parts[0];
        $expectedPacked = $parts[1];

        try {
            $gotPacked = MPCDesignation::convertSimple($unpacked);
        } catch (MPCDesignationException $e) {
            $failed++;
            if (count($errors) < $maxErrors) {
                $errors[] = [$unpacked, "ERROR: " . $e->getMessage(), $expectedPacked];
            }
            continue;
        }

        if ($gotPacked !== $expectedPacked) {
            $failed++;
            if (count($errors) < $maxErrors) {
                $errors[] = [$unpacked, $gotPacked, $expectedPacked];
            }
        } else {
            $passed++;
        }

        // Progress indicator every 100,000 entries
        if ($total % 100000 === 0) {
            echo "Processed $total entries...\n";
        }
    }

    fclose($fp);

    $elapsed = microtime(true) - $startTime;

    echo "\n";
    echo "=== Test Results ===\n";
    echo "Total:  $total\n";
    echo "Passed: $passed\n";
    echo "Failed: $failed\n";
    $elapsedMs = $elapsed * 1000;
    $rate = $total / $elapsed;
    echo sprintf("Time:   %.0fms (%.1f entries/sec)\n", $elapsedMs, $rate);
    echo "\n";

    if ($failed > 0) {
        echo sprintf("=== First %d failures ===\n", count($errors));
        echo sprintf("%-25s %-15s %-15s\n", 'Input', 'Got', 'Expected');
        echo str_repeat('-', 60) . "\n";
        foreach ($errors as [$inp, $got, $expected]) {
            echo sprintf("%-25s %-15s %-15s\n", $inp, $got, $expected);
        }
    }

    return $failed === 0;
}

function main(array $argv): int {
    if (count($argv) < 2) {
        fwrite(STDERR, "Usage: php test_csv.php <csv_file> [max_errors]\n");
        return 1;
    }

    $csvFile = $argv[1];
    $maxErrors = isset($argv[2]) ? (int)$argv[2] : 100;

    $success = runTests($csvFile, $maxErrors);
    return $success ? 0 : 1;
}

exit(main($argv));
